File: 04_Connected_Areas_in_Matrix.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<bool>> visited;
vector<pair<pair<int, int>, int>> areas;
int rows, cols;
vector<string> matrix;

int explore_area(int row, int col) {
    if (row < 0 || row >= rows || col < 0 || col >= cols) {
        return 0;
    }
    if (matrix[row][col] != '-' || visited[row][col]) {
        return 0;
    }

    visited[row][col] = true;
    int size = 1;
    size += explore_area(row - 1, col);  // explore up
    size += explore_area(row + 1, col);  // explore down
    size += explore_area(row, col - 1);  // explore left
    size += explore_area(row, col + 1);  // explore right
    return size;
}

bool compareAreas(pair<pair<int, int>, int> a, pair<pair<int, int>, int> b) {
    if (a.second != b.second) {
        return a.second > b.second;
    }
    if (a.first.first != b.first.first) {
        return a.first.first < b.first.first;
    }
    return a.first.second < b.first.second;
}

int main() {
    cin >> rows >> cols;
    matrix.resize(rows);
    visited.resize(rows, vector<bool>(cols, false));

    for (int i = 0; i < rows; i++) {
        cin >> matrix[i];
    }

    for (int row = 0; row < rows; row++) {
        for (int col = 0; col < cols; col++) {
            if (matrix[row][col] == '-' && !visited[row][col]) {
                int area_size = explore_area(row, col);
                areas.push_back({{row, col}, area_size});
            }
        }
    }

    sort(areas.begin(), areas.end(), compareAreas);

    cout << "Total areas found: " << areas.size() << "\n";
    for (int i = 0; i < areas.size(); i++) {
        cout << "Area #" << i + 1 << " at (" << areas[i].first.first << ", " << areas[i].first.second << "), size: " << areas[i].second << "\n";
    }

    return 0;
}

File: 04_Connected_Areas_in_Matrix.py
def find_connected_areas(matrix):
    rows = len(matrix)
    cols = len(matrix[0])
    visited = [[False] * cols for _ in range(rows)]
    areas = []

    def explore_area(row, col):
        if row < 0 or row >= rows or col < 0 or col >= cols:
            return 0
        if matrix[row][col] != "-" or visited[row][col]:
            return 0

        visited[row][col] = True
        size = 1
        size += explore_area(row - 1, col)  # explore up
        size += explore_area(row + 1, col)  # explore down
        size += explore_area(row, col - 1)  # explore left
        size += explore_area(row, col + 1)  # explore right
        return size

    for row in range(rows):
        for col in range(cols):
            if matrix[row][col] == "-" and not visited[row][col]:
                area_size = explore_area(row, col)
                areas.append((row, col, area_size))

    areas.sort(key=lambda x: (-x[2], x[0], x[1]))  # Sort areas by size, position

    return areas


# Пример на използване:
matrix_rows = int(input())
matrix_cols = int(input())
matrix = []
for _ in range(matrix_rows):
    row = input()
    matrix.append(row)

connected_areas = find_connected_areas(matrix)

print(f"Total areas found: {len(connected_areas)}")
for i, area in enumerate(connected_areas, 1):
    row, col, size = area
    print(f"Area #{i} at ({row}, {col}), size: {size}")
